### PR TITLE
Feature: UPS Freight Shipments/Labels

### PR DESCRIPTION
**Pull request by mamhoff**
This adds support for generating UPS Freight labels and Bills of Lading in PDF Format.

The request we generate is very similar to, but not the same format as, the rates request for UPS Freight, and we are able to share many of the options. The request generation is sufficiently different to warrant its own class - but many of the remaining generators (location hashes etc.) could be shared.

### CODE REVIEW

File: spec/friendly_shipping/services/ups_freight/label_options_spec.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'friendly_shipping/services/ups_freight/label_options'
+
+RSpec.describe FriendlyShipping::Services::UpsFreight::RatesOptions do
```
Comment (pelargir) [inline spec/friendly_shipping/services/ups_freight/label_options_spec.rb]: Should be `LabelOptions` not `RateOptions`

File: spec/friendly_shipping/services/ups_freight/label_item_options_spec.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'friendly_shipping/services/ups_freight/label_item_options'
+
+RSpec.describe FriendlyShipping::Services::UpsFreight::LabelItemOptions do
+  subject(:rates_item_options) { described_class.new(item_id: nil) }
```
Comment (pelargir) [inline spec/friendly_shipping/services/ups_freight/label_item_options_spec.rb]: `subject(:label_item_options)`

File: spec/friendly_shipping/services/ups_freight/label_package_options_spec.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'friendly_shipping/services/ups_freight/label_package_options'
+
+RSpec.describe FriendlyShipping::Services::UpsFreight::LabelPackageOptions do
+  subject(:rates_package_options) { described_class.new(package_id: nil) }
```
Comment (pelargir) [inline spec/friendly_shipping/services/ups_freight/label_package_options_spec.rb]: `subject(:label_package_options)`

File: spec/friendly_shipping/services/ups_freight/label_options_spec.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  subject(:rates_options) do
```
Comment (pelargir) [inline spec/friendly_shipping/services/ups_freight/label_options_spec.rb]: `subject(:label_options)`

File: lib/friendly_shipping/services/ups_freight.rb [hunk 5]
```diff
@@ -62,6 +67,31 @@ def rate_estimates(shipment, options:, debug: false)
...
+            Username: login,
+            Password: password,
+            AccessLicenseNumber: key
+          },
+          body: freight_rate_request_hash.to_json,
+          debug: debug
+        )
+
+        client.post(request).bind do |response|
+          ParseFreightRateResponse.call(response: response, request: request)
```
Comment (pelargir) [inline lib/friendly_shipping/services/ups_freight.rb]: `ParseFreightLabelResponse`

File: spec/friendly_shipping/services/ups_freight_spec.rb [hunk 6]
```diff
@@ -153,4 +154,107 @@
       expect(rate.data[:days_in_transit]).to eq(2)
     end
   end
+
+  describe 'labels' do
+    subject { service.rate_estimates(shipment, options: options) }
```
Comment (pelargir) [inline spec/friendly_shipping/services/ups_freight_spec.rb]: This subject can be deleted. It's re-defined below.

File: lib/friendly_shipping/services/ups_freight/generate_email_options_hash.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module FriendlyShipping
+  module Services
+    class UpsFreight
+      class GenerateEmailOptionsHash
+        def self.call(email_options:)
+          {
+            EMailInformation: {
```
Comment (pelargir) [inline lib/friendly_shipping/services/ups_freight/generate_email_options_hash.rb]: Wow, I hate how they want this to be "EMail" and not "Email." Looks terrible!

Comment (mamhoff) [reply lib/friendly_shipping/services/ups_freight/generate_email_options_hash.rb]: CamElCaSe ftw

**Comment by pelargir** [comment]
When are you planning on merging this?

**Pull request merged**